iceoryx_hoofs/test/moduletests/test_cxx_function.cpp

```diff
@@ -0,0 +1,630 @@
+    return arg.value;
+}
+
+
+class function_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+
+    static int32_t staticFunction(int32_t n)
+    {
+        return n + 1;
+    }
+};
+
+TEST_F(function_test, DefaultConstructionCreatesNoCallable)
+{
+    test_function sut;
+
+    EXPECT_FALSE(sut.operator bool());
+}
+
+TEST_F(function_test, ConstructionFromFunctorIsCallable)
+{
+    Functor f(73);
+    Functor::resetCounts();
+    test_function sut(f);
+
+    EXPECT_EQ(Functor::numCreated, 1U);
+    ASSERT_TRUE(sut.operator bool());
+    EXPECT_EQ(sut(1), f(1));
+}
+
+TEST_F(function_test, ConstructionFromLambdaIsCallable)
+{
+    int32_t capture = 37;
+    auto lambda = [state = capture](int32_t n) { return state + n; };
+    test_function sut(lambda);
+
+    ASSERT_TRUE(sut.operator bool());
+    EXPECT_EQ(sut(1), lambda(1));
+}
+
+TEST_F(function_test, ConstructionFromFreeFunctionIsCallable)
+{
+    test_function sut(freeFunction);
+
+    ASSERT_TRUE(sut.operator bool());
+    EXPECT_EQ(sut(1), freeFunction(1));
+}
+
+TEST_F(function_test, ConstructionFromStaticFunctionIsCallable)
+{
+    // is essentially also a free function but we test the case to be sure
+    test_function sut(staticFunction);
+
+    ASSERT_TRUE(sut.operator bool());
+    EXPECT_EQ(sut(1), staticFunction(1));
+}
+
+TEST_F(function_test, ConstructionFromMemberFunctionIsCallable)
+{
+    Functor f(37);
+    test_function sut(f, &Functor::operator());
+
+    ASSERT_TRUE(sut.operator bool());
+    auto result = f(1);
+    EXPECT_EQ(sut(1), result + 1);
+}
+
+TEST_F(function_test, ConstructionFromConstMemberFunctionIsCallable)
+{
+    Functor f(37);
+    test_function sut(f, &Functor::getState);
+
+    ASSERT_TRUE(sut.operator bool());
+    auto state = f.getState(1);
+    EXPECT_EQ(sut(1), state);
+    EXPECT_EQ(f.getState(1), state); // state is unchanged by the previous call
+}
+
+TEST_F(function_test, ConstructionFromAnotherFunctionIsCallable)
+{
+    constexpr int32_t INITIAL = 37;
+    int32_t capture = INITIAL;
+    auto lambda = [&](int32_t n) { return ++capture + n; };
+    function<signature, Bytes / 2> f(lambda); // the other function type must be small enough to fit
+    test_function sut(f);
+
+    ASSERT_TRUE(sut.operator bool());
+    auto result = f(1);
+    EXPECT_EQ(sut(1), result + 1);
+    EXPECT_EQ(capture, INITIAL + 2);
+}
+
+TEST_F(function_test, FunctionStateIsIndependentOfSource)
+{
+    constexpr uint32_t INITIAL_STATE = 73U;
+    static_storage<1024U> storage;
+    auto p = storage.allocate<Functor>();
+    p = new (p) Functor(INITIAL_STATE);
+    auto& functor = *p;
+
+    // test whether the function really owns the functor
+    // (no dependency or side effects)
+    test_function sut(functor);
+
+    ASSERT_TRUE(sut.operator bool());
+
+    // both increment their state independently
+    EXPECT_EQ(sut(1U), functor(1U));
+
+    // clear original (set to 0)
+    p->~Functor();
```

Comment:
Please don't call the d'tor directly in tests. Either use an `cxx::optional` or a `std::unique_ptr` to wrap the object.

Reply:
This is needed as it was constructed using placement new and is the way to go. This is not a use case for an `optional`. We cannot really leak anything here (if we leave via ASSERT), but we can prevent the missed call with a `unique_ptr'` acting as guard.